xds/internal/resolver/serviceconfig_action.go

```diff
@@ -0,0 +1,158 @@
+	// The assigned name, clusters plus index number, "A_B_1"
+	assignedName string
+}
+
+// oldActions contains the final actions. The following code updates
+// oldActions with updates from RDS resp.
+// var oldActions map[string]actionsWithNextIndex
+// var oldActions map[string]map[string]action
+
+// Turn an RDS response into a map of actions.
+// newActions contains new actions, with names unassigned
+//
+// Note that this only handles actions, but not matches. The routes in the
+// generated balancer config should maintain the same order as in RDS resp.
+// A separate list of routes would work.
+//
+
+// newActionsFromRoutes gets actions from the routes, and turns them into a map
+// keyed by the hash of the clusters.
+//
+// In the returned map, all actions don't have assignedName. The assignedName
+// will be filled in after comparing the new actions with the existing actions,
+// so when a new and old action only diff in weights, the new action can reuse
+// the old action's name.
+//
+// from
+//   {B:60, A:40}, {A:30, B:70}, {B:90, C:10}
+//
+// to
+//   A40_B60_: {{A:40, B:60}, "A_B_", ""}
+//   A30_B70_: {{A:30, B:70}, "A_B_", ""}
+//   B90_C10_: {{B:90, C:10}, "B_C_", ""}
+func newActionsFromRoutes(routes []*xdsclient.Route) map[string]action {
+	newActions := make(map[string]action)
+
+	for _, route := range routes {
+		var clusterNames []string
+		clusters := route.Action
+		for n := range clusters {
+			clusterNames = append(clusterNames, n)
+		}
+
+		// Sort names to be consistent.
+		sort.Strings(clusterNames)
+		clustersOnly := ""
+		clustersWithWeight := ""
+		for _, c := range clusterNames {
+			// Generates A_B_
+			clustersOnly = clustersOnly + c + "_"
+			// Generates A40_B60_
+			clustersWithWeight = clustersWithWeight + c + strconv.FormatUint(uint64(clusters[c]), 10) + "_"
+		}
+
+		if _, ok := newActions[clustersWithWeight]; !ok {
+			newActions[clustersWithWeight] = action{
+				clustersWithWeights: clusters,
+				clusterNames:        clustersOnly,
+			}
+		}
+	}
+	return newActions
+}
+
+// updateActions takes a new map of actions, and updates the existing action map in the resolver.
+//
+// In the old map, all actions have assignedName set.
+// In the new map, all actions have no assignedName.
+//
+// After the update, the action map is updated to have all actions from the new
+// map, with assignedName:
+// - if the new action exists in old, get the old name
+// - if the new action doesn't exist in old
+//   - if there is an old action that will be removed, and has the same set of
+//   clusters, reuse the old action's name
+//   - otherwise, generate a new name
+func (r *xdsResolver) updateActions(newActions map[string]action) {
+	// Delete actions from oldActions if they are not in newActions.
+	// Keep the removed actions in a map, with key as clusterNames without
+	// weights, so their assigned names can be reused.
+	if r.actions == nil {
+		r.actions = make(map[string]action)
+	}
+	oldActions := r.actions
```

Comment:
The name `oldActions` gives the feeling that this would be discarded. But it looks like the new actions are added to this. So, maybe name is `existingActions`?

Reply:
Done